File: src/main/java/net/flyingfishflash/loremlist/persistence/LrmListItemRepositoryRdbms.java
package net.flyingfishflash.loremlist.persistence;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import kotlin.Pair;
import kotlinx.datetime.Clock;
import kotlinx.datetime.ConvertersKt;
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemSuccinct;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItem;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemRepository;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

@Repository
class LrmListItemRepositoryRdbms implements LrmListItemRepository {

  private final JdbcClient jdbcClient;

  LrmListItemRepositoryRdbms(JdbcClient jdbcClient) {
    this.jdbcClient = jdbcClient;
  }

  @Override
  public long countByOwnerAndListId(UUID listId, String listOwner) {
    return jdbcClient
        .sql(
            "SELECT COUNT(*) FROM list_item li JOIN list l ON l.id = li.list_id "
                + "WHERE li.list_id = :listId AND l.owner = :listOwner")
        .param("listId", listId)
        .param("listOwner", listOwner)
        .query(Long.class)
        .single();
  }

  @Override
  public long countByOwnerAndItemId(UUID itemId, String itemOwner) {
    return jdbcClient
        .sql(
            "SELECT COUNT(*) FROM list_item li JOIN item i ON i.id = li.item_id "
                + "WHERE li.item_id = :itemId AND i.owner = :itemOwner")
        .param("itemId", itemId)
        .param("itemOwner", itemOwner)
        .query(Long.class)
        .single();
  }

  @Override
  public void create(UUID listId, UUID itemId) {
    jdbcClient
        .sql(
            "INSERT INTO list_item (list_id, item_id, item_quantity, item_is_suppressed) "
                + "VALUES (:listId, :itemId, 0, false)")
        .param("listId", listId)
        .param("itemId", itemId)
        .update();
  }

  @Override
  public List<SuccinctLrmComponentPair> create(Set<Pair<UUID, UUID>> associationCollection) {
    if (associationCollection.isEmpty()) {
      return List.of();
    }

    for (Pair<UUID, UUID> association : associationCollection) {
      create(association.getFirst(), association.getSecond());
    }

    Set<UUID> listIds =
        associationCollection.stream().map(Pair::getFirst).collect(Collectors.toSet());
    Set<UUID> itemIds =
        associationCollection.stream().map(Pair::getSecond).collect(Collectors.toSet());

    Map<UUID, String> listNamesById =
        jdbcClient
            .sql("SELECT id, name FROM list WHERE id IN (:ids)")
            .param("ids", listIds)
            .query((rs, rowNum) -> Map.entry((UUID) rs.getObject("id"), rs.getString("name")))
            .list()
            .stream()
            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

    Map<UUID, String> itemNamesById =
        jdbcClient
            .sql("SELECT id, name FROM item WHERE id IN (:ids)")
            .param("ids", itemIds)
            .query((rs, rowNum) -> Map.entry((UUID) rs.getObject("id"), rs.getString("name")))
            .list()
            .stream()
            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

    return associationCollection.stream()
        .map(
            association ->
                new SuccinctLrmComponentPair(
                    new LrmListSuccinct(
                        association.getFirst(), listNamesById.get(association.getFirst())),
                    new LrmItemSuccinct(
                        association.getSecond(), itemNamesById.get(association.getSecond()))))
        .toList();
  }

  @Override
  public LrmListItem findByOwnerAndItemIdAndListIdOrNull(UUID itemId, UUID listId, String owner) {
    return jdbcClient
        .sql(
            "SELECT i.id AS item_id, i.name AS item_name, i.description AS item_description, "
                + "i.owner AS item_owner, i.created AS item_created, i.creator AS item_creator, "
                + "i.updated AS item_updated, i.updater AS item_updater, "
                + "li.list_id AS li_list_id, li.item_quantity AS li_item_quantity, "
                + "li.item_is_suppressed AS li_item_is_suppressed, l.name AS list_name "
                + "FROM item i "
                + "JOIN list_item li ON li.item_id = i.id "
                + "JOIN list l ON l.id = li.list_id "
                + "WHERE i.id = :itemId AND li.list_id = :listId AND i.owner = :owner AND l.owner = :owner")
        .param("itemId", itemId)
        .param("listId", listId)
        .param("owner", owner)
        .query(
            (rs, rowNum) ->
                new LrmListItem(
                    (UUID) rs.getObject("item_id"),
                    (UUID) rs.getObject("li_list_id"),
                    rs.getString("item_name"),
                    rs.getString("item_description"),
                    rs.getInt("li_item_quantity"),
                    rs.getBoolean("li_item_is_suppressed"),
                    rs.getString("item_owner"),
                    new kotlinx.datetime.Instant(rs.getObject("item_created", Instant.class)),
                    rs.getString("item_creator"),
                    new kotlinx.datetime.Instant(rs.getObject("item_updated", Instant.class)),
                    rs.getString("item_updater"),
                    Set.of(
                        new LrmListSuccinct(
                            (UUID) rs.getObject("li_list_id"), rs.getString("list_name")))))
        .optional()
        .orElse(null);
  }

  // TODO: add owner restriction
  @Override
  public int removeByOwnerAndItemId(UUID itemId, String owner) {
    return jdbcClient
        .sql("DELETE FROM list_item WHERE item_id = :itemId")
        .param("itemId", itemId)
        .update();
  }

  // TODO: add owner restriction
  @Override
  public int removeByOwnerAndListId(UUID listId, String owner) {
    return jdbcClient
        .sql("DELETE FROM list_item WHERE list_id = :listId")
        .param("listId", listId)
        .update();
  }

  // TODO: add owner restriction
  @Override
  public int removeByOwnerAndListIdAndItemId(UUID listId, UUID itemId, String owner) {
    int deletedCount =
        jdbcClient
            .sql("DELETE FROM list_item WHERE list_id = :listId AND item_id = :itemId")
            .param("listId", listId)
            .param("itemId", itemId)
            .update();
    if (deletedCount > 1) {
      throw new IllegalStateException("This query should delete no more than one record.");
    }
    return deletedCount;
  }

  @Override
  public int updateName(LrmListItem lrmListItem) {
    return jdbcClient
        .sql("UPDATE item SET name = :name, updated = :updated WHERE id = :id")
        .param("name", lrmListItem.name())
        .param("updated", toJavaInstant(Clock.System.INSTANCE.now()))
        .param("id", lrmListItem.id())
        .update();
  }

  @Override
  public int updateDescription(LrmListItem lrmListItem) {
    return jdbcClient
        .sql("UPDATE item SET description = :description, updated = :updated WHERE id = :id")
        .param("description", lrmListItem.description())
        .param("updated", toJavaInstant(Clock.System.INSTANCE.now()))
        .param("id", lrmListItem.id())
        .update();
  }

  @Override
  public int updateQuantity(LrmListItem lrmListItem) {
    int updatedCount =
        jdbcClient
            .sql(
                "UPDATE list_item SET item_quantity = :quantity WHERE item_id = :itemId AND list_id = :listId")
            .param("quantity", lrmListItem.quantity())
            .param("itemId", lrmListItem.id())
            .param("listId", lrmListItem.listId())
            .update();
    if (updatedCount > 1) {
      throw new IllegalStateException("This query should update no more than one record.");
    }
    return updatedCount;
  }

  @Override
  public int updateIsItemSuppressed(LrmListItem lrmListItem) {
    int updatedCount =
        jdbcClient
            .sql(
                "UPDATE list_item SET item_is_suppressed = :isSuppressed "
                    + "WHERE item_id = :itemId AND list_id = :listId")
            .param("isSuppressed", lrmListItem.isSuppressed())
            .param("itemId", lrmListItem.id())
            .param("listId", lrmListItem.listId())
            .update();
    if (updatedCount > 1) {
      throw new IllegalStateException("This query should update no more than one record.");
    }
    return updatedCount;
  }

  @Override
  public int updateListId(LrmListItem lrmListItem, UUID destinationListId) {
    int updatedCount =
        jdbcClient
            .sql(
                "UPDATE list_item SET list_id = :destinationListId WHERE item_id = :itemId AND list_id = :listId")
            .param("destinationListId", destinationListId)
            .param("itemId", lrmListItem.id())
            .param("listId", lrmListItem.listId())
            .update();
    if (updatedCount != 1) {
      throw new IllegalStateException("This query should update exactly one record.");
    }
    return updatedCount;
  }

  private static Instant toJavaInstant(kotlinx.datetime.Instant instant) {
    return ConvertersKt.toJavaInstant(instant);
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/persistence/LrmListItemRepositoryRdbms.kt
package net.flyingfishflash.loremlist.persistence

import kotlinx.datetime.Clock.System.now
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemSuccinct
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItem
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemRepository
import net.flyingfishflash.loremlist.persistence.LrmListItemsTable.item
import net.flyingfishflash.loremlist.persistence.LrmListItemsTable.itemIsSuppressed
import net.flyingfishflash.loremlist.persistence.LrmListItemsTable.itemQuantity
import net.flyingfishflash.loremlist.persistence.LrmListItemsTable.list
import org.jetbrains.exposed.sql.Query
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.andWhere
import org.jetbrains.exposed.sql.batchInsert
import org.jetbrains.exposed.sql.count
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.statements.UpdateStatement
import org.jetbrains.exposed.sql.update
import org.springframework.stereotype.Repository
import java.util.*

@Repository
class LrmListItemRepositoryRdbms : LrmListItemRepository {

  override fun countByOwnerAndListId(listId: UUID, listOwner: String): Long {
    val listCount = list.count()
    return (LrmListItemsTable leftJoin LrmListTable).select(listCount)
      .where { list eq listId }
      .andWhere { LrmListTable.owner eq listOwner }.map { it[listCount] }.first()
  }

  override fun countByOwnerAndItemId(itemId: UUID, itemOwner: String): Long {
    val itemCount = item.count()
    return (LrmListItemsTable leftJoin LrmItemTable).select(itemCount)
      .where { item eq itemId }
      .andWhere { LrmItemTable.owner eq itemOwner }.map { it[itemCount] }.first()
  }

  override fun create(listId: UUID, itemId: UUID) {
    LrmListItemsTable.insert {
      it[list] = listId
      it[item] = itemId
      it[itemQuantity] = 0
      it[itemIsSuppressed] = false
    }
  }

  // TODO: create custom type for associationCollection
  override fun create(associationCollection: Set<Pair<UUID, UUID>>): List<SuccinctLrmComponentPair> {
    val associationIdSet = LrmListItemsTable.batchInsert(associationCollection) { (listId, itemId) ->
      this[list] = listId
      this[item] = itemId
      this[itemQuantity] = 0
      this[itemIsSuppressed] = false
    }.map { Pair(it[list], it[item]) }.toSet()

    val succinctLrmComponentPairs = (LrmListItemsTable innerJoin LrmListTable innerJoin LrmItemTable)
      .select(LrmListTable.id, LrmListTable.name, LrmItemTable.id, LrmItemTable.name)
      .where { Pair(list, item) inList (associationIdSet) }
      .map {
        SuccinctLrmComponentPair(
          list = LrmListSuccinct(it[LrmListTable.id], it[LrmListTable.name]),
          item = LrmItemSuccinct(it[LrmItemTable.id], it[LrmItemTable.name]),
        )
      }

    return succinctLrmComponentPairs
  }

  override fun findByOwnerAndItemIdAndListIdOrNull(itemId: UUID, listId: UUID, owner: String): LrmListItem? {
    val lrmListItems = mapQueryResultToLrmItems(
      (LrmItemTable leftJoin LrmListItemsTable leftJoin LrmListTable)
        .selectAll()
        .where { item eq itemId }
        .andWhere { list eq listId }
        .andWhere { LrmItemTable.owner eq owner }
        .andWhere { LrmListTable.owner eq owner },
    ).distinct()
    check(lrmListItems.size <= 1) { "This query should return no more than one record." }
    return lrmListItems.firstOrNull()
  }

  // TODO: add owner restriction
  override fun removeByOwnerAndItemId(itemId: UUID, owner: String): Int {
    return LrmListItemsTable.deleteWhere { item eq itemId }
  }

  // TODO: add owner restriction
  override fun removeByOwnerAndListId(listId: UUID, owner: String): Int {
    return LrmListItemsTable.deleteWhere { list eq listId }
  }

  // TODO: add owner restriction
  override fun removeByOwnerAndListIdAndItemId(listId: UUID, itemId: UUID, owner: String): Int {
    val deletedCount = LrmListItemsTable.deleteWhere { list eq listId and (item eq itemId) }
    check(deletedCount <= 1) { "This query should delete no more than one record." }
    return deletedCount
  }

  override fun updateName(lrmListItem: LrmListItem): Int = updateItem(lrmListItem) { it[LrmItemTable.name] = lrmListItem.name }

  override fun updateDescription(lrmListItem: LrmListItem): Int = updateItem(lrmListItem) {
    it[LrmItemTable.description] =
      lrmListItem.description
  }

  override fun updateQuantity(lrmListItem: LrmListItem): Int {
    val updatedCount = LrmListItemsTable.update({ byListIdAndItemId(lrmListItem) }) { it[itemQuantity] = lrmListItem.quantity }
    check(updatedCount <= 1) { "This query should update no more than one record." }
    return updatedCount
  }

  override fun updateIsItemSuppressed(lrmListItem: LrmListItem): Int {
    val updatedCount = LrmListItemsTable.update({ byListIdAndItemId(lrmListItem) }) { it[itemIsSuppressed] = lrmListItem.isSuppressed }
    check(updatedCount <= 1) { "This query should update no more than one record." }
    return updatedCount
  }

  override fun updateListId(lrmListItem: LrmListItem, destinationListId: UUID): Int {
    val updatedCount = LrmListItemsTable.update({ byListIdAndItemId(lrmListItem) }) { it[list] = destinationListId }
    check(updatedCount == 1) { "This query should update exactly one record." }
    return updatedCount
  }

  private fun updateItem(lrmItem: LrmListItem, updateAction: (UpdateStatement) -> Unit): Int {
    return LrmItemTable.update({ LrmItemTable.id eq lrmItem.id }) {
      updateAction(it)
      it[updated] = now()
    }
  }

  private fun mapQueryResultToLrmItems(query: Query): List<LrmListItem> {
    val resultRows = query.toList()

    @Suppress("SENSELESS_COMPARISON")
    val listsByItems = resultRows
      .filter { it[LrmListTable.id] != null }
      .groupBy(
        keySelector = { it[LrmItemTable.id] },
        valueTransform = {
          LrmListSuccinct(
            id = it[LrmListTable.id],
            name = it[LrmListTable.name],
          )
        },
      )

    return resultRows
      .distinctBy { it[LrmItemTable.id] }
      .map {
        it.toLrmListItem().copy(
          lists = listsByItems[it[LrmItemTable.id]]?.sortedBy { list -> list.name }?.toSet() ?: emptySet(),
        )
      }
  }

  private fun byListIdAndItemId(lrmListItem: LrmListItem) = item eq lrmListItem.id and (list eq lrmListItem.listId)

  private fun ResultRow.toLrmListItem(): LrmListItem {
    return LrmListItem(
      id = this[item],
      listId = this[list],
      name = this[LrmItemTable.name],
      description = this[LrmItemTable.description],
      owner = this[LrmItemTable.owner],
      created = this[LrmItemTable.created],
      creator = this[LrmItemTable.creator],
      updated = this[LrmItemTable.updated],
      updater = this[LrmItemTable.updater],
      quantity = this[itemQuantity],
      isSuppressed = this[itemIsSuppressed],
    )
  }
}
